### PR TITLE
Add `#![no_std]` support

### PR DESCRIPTION
**Pull request by overcat**
### What

add [`#![no_std]`](https://docs.rust-embedded.org/book/intro/no-std.html#summary) support.

### Why

Allow it to work on embedded devices (such as STM32).

### Known limitations

N/A

### CODE REVIEW

File: Cargo.toml
Comment (overcat) [inline Cargo.toml:31]: The `thiserror` crate is rarely used in this project, and it does not support `no_std`, so I choose to remove it.

Comment (leighmcculloch) [reply Cargo.toml:31]: I think this is the right call. 👍🏻

File: Cargo.toml
Comment (leighmcculloch) [inline Cargo.toml:31]: This might be a hard question to answer without some experimentation, but do we need alloc? All of the inputs and outputs of the strkey library are constrained in size. We _should_ I think be able to do all the base32 enc/dec without allocating, since the sizes should all be relatively small and predictable. Wdyt?

**Review by leighmcculloch — COMMENTED** [review]
Can we explore making the crate noalloc? I think it makes for a much bigger case for supporting nostd and was what we were hoping to do. I think because of the functionality in the crate, that the size of what is being enc/dec is always predictable, this should be possible. Although I don't know how well the data-encoding crate is at supporting that. Wdyt?

**Comment by overcat** [comment]
Hi @leighmcculloch, I have initially added support for no alloc, the current implementation is quite rough, and I will improve it tomorrow. It would be great if you could review the interface, but please do not focus on the details of the implementation at this time.

File: .github/workflows/rust.yml
Comment (leighmcculloch) [inline .github/workflows/rust.yml:59]: On other repos where we want to test a variety of features, we typically use cargo-hack. It does a good job of testing feature combinations. It'll test the default case too.
```suggestion
    - uses: stellar/binaries@v30
      with:
        name: cargo-hack
        version: 0.5.28
    - run: cargo-hack hack --feature-powerset --target ${{ matrix.sys.target }}
```

**Review by leighmcculloch — COMMENTED** [review]
@overcat Thanks for taking a look at what it would take to make this go no-alloc and for the time you put into it ❤️.

In interest of getting the no-std functionality you need merged asap you should revert the last few no-alloc commits, because looking back at it the no-std change is on its own much more surgical and easier to review, and makes no breaking changes to the API.

The no-alloc change as proposed requires breaking changes to the API that I'd like some more time to think about, but I don't want to delay getting a no-std version out.

Thanks for persevering with me on this, I didn't realize scope difference of the no-alloc vs no-std approach.

**Comment by leighmcculloch** [comment]
When you revert the no-alloc commits, could you do so with a revert, so it's easy for me to come back and extract the no-alloc commits to experiment with further with the API?

File: Cargo.toml
Comment (leighmcculloch) [inline Cargo.toml:16]: It just occurred to me, that we don't need the `std` feature.

Nothing in this change requires `libstd` as everything is in `libcore`. So the lib can be `nostd`, and anyone can import it into std or nostd projects.

Comment (overcat) [reply Cargo.toml:16]: Yes, I will modify it after the merge of #70.

File: README.md
Comment (leighmcculloch) [inline README.md:19]: I don't think users will need to specifically enable the alloc crate? The fact that this crate extern references the alloc crate makes it required. Am I missing something?
```suggestion
By default, the library uses the standard library (std). If you want to use it in a `no_std` environment, you need to disable the `default` features.
```

Comment (overcat) [reply README.md:19]: Yes, there is an issue with the wording here. I want to mention that this library uses `alloc` because when using this library in a `no_std` environment, due to the absence of a default global memory allocator, users need to add one themselves.

Perhaps it would be better to modify the expression as follows?

> This crate can be used in `no_std` environments. However, please note that it relies on the `alloc` crate for certain types such as `Vec`. When using this library in a `no_std` context, you are required to provide a global memory allocator. You can define a global allocator using the [`#[global_allocator]`](https://doc.rust-lang.org/std/alloc/index.html#the-global_allocator-attribute) attribute.

Comment (leighmcculloch) [reply README.md:19]: I like it.

I think it would also be reasonable to link out to instructions for how to use the `alloc` crate, instead of providing them here. For example, this page shows how to use the alloc crate: https://docs.rust-embedded.org/book/collections/

But I don't feel strongly one way or the other. 👍🏻

**Comment by leighmcculloch** [comment]
@overcat Can you merge `main` into this PR?

**Comment by overcat** [comment]
Hi @leighmcculloch, the current CI is failing, I think it's due to the Action cache, but I don't have permission to clear the cache.

File: .github/workflows/rust.yml
Comment (leighmcculloch) [inline .github/workflows/rust.yml:54]: I think we can revert this part of the change now that no additional features are added.

File: Cargo.toml
Comment (leighmcculloch) [inline Cargo.toml:23]: ❤️

File: README.md
Comment (leighmcculloch) [inline README.md:19]:
```suggestion
This crate does not depend on the `std` crate and can be used in `no_std` environments. 

```

File: tests/tests.rs
Comment (leighmcculloch) [inline tests/tests.rs:1]: Tests always run with std imported anyway, because the test framework itself uses std, so I don't think we need the changes in this file.

Is there a reason using alloc explicitly here benefits?

**Review by leighmcculloch — COMMENTED** [review]
Two small changes requested inline, otherwise this looks great.

File: .github/workflows/rust.yml
Comment (overcat) [reply .github/workflows/rust.yml:54]: I think we can keep the current way, the previous testing method will not test whether cli can be compiled normally.

Comment (leighmcculloch) [reply .github/workflows/rust.yml:54]: Ah good call. 👍🏻 💯 ❤️

**Review by leighmcculloch — APPROVED** [review]

File: tests/tests.rs
Comment (leighmcculloch) [reply tests/tests.rs]: We could add a crate to the repo that imports strkey and gets built as part of testing, that provides it's own panic handler, because the only way it'll build is if std isn't imported because there can't be two panic handler and std provides one. 

https://doc.rust-lang.org/nomicon/panic-handler.html

Comment (leighmcculloch) [reply tests/tests.rs]: On the same note that's a lot of test machinery for checking something that is unlikely to be the case since the no_std attribute is in use.

The risk is that it depends on a crate that pulls in std.

I wonder if something like https://github.com/cackle-rs/cackle could detect it.

Comment (overcat) [reply tests/tests.rs]: Hi @leighmcculloch, I have integrated version 0.0.11 into Ledger, and it's working very well. Currently, I want to make rs-stellar-xdr support no std, so I will work on the tests here and consider adding no alloc support in my free time.

Comment (leighmcculloch) [reply tests/tests.rs]: There's nothing riding on no alloc, so if it's not of value to you, I suggest holding off as there's nothing htat I know of needing it right now.

**Pull request merged**
